pkg/analysis/passes/checksum/checksum.go

```diff
@@ -0,0 +1,155 @@
+)
+
+var Analyzer = &analysis.Analyzer{
+	Name:     "checksum",
+	Requires: []*analysis.Analyzer{readme.Analyzer},
+	Run:      run,
+	Rules:    []*analysis.Rule{checksumInvalid},
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+
+	checksum := pass.CheckParams.Checksum
+
+	// skip if no provided
+	if checksum == "" || len(checksum) == 0 {
+		return nil, nil
+	}
+
+	checksum, err := resolveCheckSum(strings.TrimSpace(checksum))
+	if err != nil {
+		pass.ReportResult(
+			pass.AnalyzerName,
+			checksumFormatError,
+			fmt.Sprintf("%s: %s", err.Error(), pass.CheckParams.Checksum),
+			"Make sure you provide a valid checksum as a string or a valid URL to a checksum. Only MD5 and SHA1 checksums are supported.",
+		)
+		return nil, nil
+	}
+
+	checkSumType := ""
+
+	// determine if it is md5 or sha1
+	if len(checksum) == 32 {
+		checkSumType = "md5"
+	} else if len(checksum) == 40 {
+		checkSumType = "sha1"
+	} else {
+		pass.ReportResult(
+			pass.AnalyzerName,
+			checksumFormatError,
+			fmt.Sprintf("Invalid checksum format: %s", checksum),
+			"Make sure you provide a valid checksum as a string or a valid URL to a checksum. Only MD5 and SHA1 checksums are supported.",
+		)
+		return nil, nil
+	}
+	isError := false
+
+	switch checkSumType {
+	case "md5":
+		if checksum != pass.CheckParams.ArchiveCalculatedMD5 {
+			isError = true
+		}
+	case "sha1":
+		if checksum != pass.CheckParams.ArchiveCalculatedSHA1 {
+			isError = true
+		}
+	}
+
+	if isError {
+		pass.ReportResult(
+			pass.AnalyzerName,
+			checksumInvalid,
+			fmt.Sprintf("The provided checksum %s does not match the plugin archive", checksum),
+			"The plugin archive does not match the provided checksum in the submission form. Please double check you are providing the correct plugin archive and checksum.",
+		)
+	}
+
+	return nil, nil
+}
+
+// sums can come as:
+// 04b2f189616b65985d219ca094f3609b  arduino-cli.yaml
+// 0b41e5bffba89764a494f420347fe5110fa2886c  arduino-cli.yaml
+// 0b41e5bffba89764a494f420347fe5110fa2886c
+// 04b2f189616b65985d219ca094f3609b
+// we only care about the first part
+func sanitizeCheckSum(checksum string) string {
+	return strings.Split(checksum, " ")[0]
+}
+
+// checksum can be urls, we need to download
+func resolveCheckSum(checksum string) (string, error) {
+
+	if strings.HasPrefix(checksum, "https://") || strings.HasPrefix(checksum, "http://") {
+		finalChecksum, err := getFirstLineFromURL(checksum)
+		if err != nil {
+			return "", err
+		}
+		return sanitizeCheckSum(finalChecksum), nil
+	}
+	return sanitizeCheckSum(checksum), nil
+}
+
+const (
+	maxSize = 500 * 1024 // 500KB
+	timeout = 30 * time.Second
+)
+
+func getFirstLineFromURL(url string) (string, error) {
+	client := http.Client{
+		Timeout: timeout,
+	}
+
+	resp, err := client.Get(url)
+	if err != nil {
+		logme.DebugFln("Error reading body: %s", err.Error())
+		return "", err
+	}
+
+	if resp.StatusCode != 200 {
+		return "", errors.New("checksum file not found")
```

Comment:
Status code != 200 could also be another reason, not necessarily file not found. I would also add the original status code to the error, something like:

```suggestion
		return "", fmt.Errorf("bad status code: %d", resp.StatusCode)
```

Reply:
@xnyo it is indeed. but this is an error message that the user will read (in the submission process) not to log somewhere for us so I decided to make it somewhat more user-readable and if it comes to it we can debug the issue further if it ever happens.

Reply:
Okay, sounds good then!